src/utils/logging.c

```diff
@@ -50,3 +52,27 @@ void bd_utils_log (gint level, const gchar *msg) {
     if (log_func)
         log_func (level, msg);
 }
+
+/**
+ * bd_utils_log_format:
+ * @level: log level
+ * @format: printf-style format for the log message
+ */
+void bd_utils_log_format (gint level, const gchar *format, ...) {
+    gchar *msg = NULL;
+    va_list args;
+    gint ret = 0;
+
+    if (log_func) {
+        va_start (args, format);
+        ret = g_vasprintf (&msg, format, args);
+        va_end (args);
+
+        if (ret < 0)
+            return;
+
+        log_func (level, msg);
+    }
```

Comment:
Can we do `else if (level >= WARNING) g_warning(format, args)` here so that if the user code doesn't have the logging function set, they would still get the important messages somehow?

Reply:
It's library, in most cases user won't see stdout anyway.

I was thinking about adding a logging function that logs using g_warning that users can use if they don't want to write custom logging function. Maybe we could add this and make it default logging function instead of having none?

Reply:
> It's library, in most cases user won't see stdout anyway.

Well, stdout from daemons started as systemd services by default goes to journal, doesn't it?

Reply:
> I was thinking about adding a logging function that logs using g_warning that users can use if they don't want to write custom logging function. Maybe we could add this and make it default logging function instead of having none?

Sounds like a plan to me!

Reply:
I've added a new functions that logs using g_warning/info/etc. I'm still not sure if using this function as default is a good thing.